File: sudoku-game/game-boot/src/main/java/com/zumin/sudoku/game/utils/SudokuBuilder.java
package com.zumin.sudoku.game.utils;

import com.zumin.sudoku.common.core.utils.PublicUtils;
import com.zumin.sudoku.game.pojo.bo.GameDataBO;
import java.util.Arrays;
import java.util.stream.IntStream;
import lombok.experimental.UtilityClass;

/**
 * 数独游戏构建类
 */
@UtilityClass
public class SudokuBuilder {

  /**
   * 产生随机数组时的阈值
   */
  private final int MAX_CALL_RANDOM_ARRAY_TIMES = 220;

  /**
   * 生成数独终盘
   *
   * @param minEmpty 最小的空缺格子数
   * @param maxEmpty 最大的空缺格子数
   * @return 数独数据
   */
  public GameDataBO generateSudokuFinal(int minEmpty, int maxEmpty) {
    GameDataBO sudokuData;
    do {
      sudokuData = new GameDataBO();
      generateMatrix(sudokuData.getMatrix());
      digHolesByGameDifficulty(sudokuData.getHoles(), minEmpty, maxEmpty);
    } while (SudokuUtils.hasNotOnlyOneSolution(sudokuData.getMatrix(), sudokuData.getHoles()));
    return sudokuData;
  }

  /**
   * 生成数独矩阵
   *
   * @param matrix 数独矩阵
   */
  private void generateMatrix(int[][] matrix) {
    //记录buildRandomArray()调用次数
    int currentTimes = 0;
    GenerateStatus status = GenerateStatus.INIT_FIRST_ROW;
    int row = 0, col = 0;
    int[] tempRandomArray = null;
    while (row != 9) {
      switch (status) {
        case FILL_GRID:
          if (currentTimes >= MAX_CALL_RANDOM_ARRAY_TIMES) {
            status = GenerateStatus.EMPTY_ALL;
            break;
          }
          if (!isCandidateNmbFound(matrix, tempRandomArray, row, col)) {
            status = GenerateStatus.EMPTY_ROW;
            break;
          }
          if (++col == 9) {
            row++;
            status = GenerateStatus.FILL_ROW;
          }
          break;
        case EMPTY_ROW:
          clearMatrixRow(matrix, row);
          status = GenerateStatus.FILL_ROW;
          break;
        case FILL_ROW:
          tempRandomArray = buildRandomArray();
          currentTimes++;
          col = 0;
          status = GenerateStatus.FILL_GRID;
          break;
        case INIT_FIRST_ROW:
          matrix[0] = buildRandomArray();
          currentTimes = 1;
          row = 0;
          status = GenerateStatus.FILL_ROW;
          break;
        case EMPTY_ALL:
          clearMatrix(matrix);
          status = GenerateStatus.INIT_FIRST_ROW;
          break;
      }
    }
  }

  /**
   * 清空数独矩阵的行
   *
   * @param matrix 数独矩阵
   * @param row    行
   */
  private void clearMatrixRow(int[][] matrix, int row) {
    Arrays.fill(matrix[row], 0, 9, 0);
  }

  /**
   * 清空数独矩阵
   *
   * @param matrix 数独矩阵
   */
  private void clearMatrix(int[][] matrix) {
    IntStream.range(0, 9).forEach(i -> clearMatrixRow(matrix, i));
  }

  /**
   * 尝试给数独矩阵的第row行第col列的数据赋值
   *
   * @param matrix      数独矩阵
   * @param randomArray 随机数组
   * @param row         行
   * @param col         列
   * @return 能赋值返回true，不能赋值返回false
   */
  private boolean isCandidateNmbFound(int[][] matrix, int[] randomArray, int row, int col) {
    for (int i = 0; i < randomArray.length; i++) {
      if (randomArray[i] == -1) {
        continue;
      }
      int random = randomArray[i];
      matrix[row][col] = random;
      if (noConflict(matrix, row, col)) {
        randomArray[i] = -1;
        return true;
      }
    }
    return false;
  }

  /**
   * 判断该行、该列的数在数独矩阵中是否存在冲突
   *
   * @param matrix 数独矩阵
   * @param row    行
   * @param col    列
   * @return 存在冲突返回false，不存在冲突返回true
   */
  private boolean noConflict(int[][] matrix, int row, int col) {
    return SudokuUtils.checkRowIsOnly(matrix[row], col, col) && SudokuUtils.checkColumnIsOnly(matrix, row, col, row)
        && SudokuUtils.checkBlockIsOnly(matrix, row, col);
  }

  /**
   * 返回一个1-9的随机排列数组
   *
   * @return 打乱的数组
   */
  private int[] buildRandomArray() {
    int[] array = new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9};
    PublicUtils.randomizedArray(array);
    return array;
  }

  /**
   * 根据空缺格子数生成对应题目的空缺数组
   *
   * @param holes    题目空缺数组
   * @param minEmpty 最小的空缺格子数
   * @param maxEmpty 最大的空缺格子数
   */
  private void digHolesByGameDifficulty(boolean[][] holes, int minEmpty, int maxEmpty) {
    for (int i = PublicUtils.getRandomInt(minEmpty, maxEmpty) - 1; i >= 0; i--) {
      holes[i / 9][i % 9] = true;
    }
    PublicUtils.randomizedArray(holes);
  }

  /**
   * 生成数独矩阵算法的状态
   */
  private enum GenerateStatus {
    /**
     * 清空矩阵的所有数据
     */
    EMPTY_ALL,
    /**
     * 清空矩阵的行数据
     */
    EMPTY_ROW,
    /**
     * 填充矩阵的格子数据
     */
    FILL_GRID,
    /**
     * 填充矩阵的行数据
     */
    FILL_ROW,
    /**
     * 初始化矩阵的第一行数据
     */
    INIT_FIRST_ROW
  }

}

File: sudoku-game/game-boot/src/main/java/com/zumin/sudoku/game/utils/SudokuBuilder.kt
package com.zumin.sudoku.game.utils

import cn.hutool.core.util.RandomUtil
import com.zumin.sudoku.common.core.utils.shuffle
import com.zumin.sudoku.game.pojo.GameDataBO
import java.util.*

// 产生随机数组时的阈值
private const val MAX_CALL_RANDOM_ARRAY_TIMES = 220

/**
 * 生成数独终盘
 *
 * @param minEmpty 最小的空缺格子数
 * @param maxEmpty 最大的空缺格子数
 * @return 数独数据
 */
fun generateSudokuFinal(minEmpty: Int, maxEmpty: Int): GameDataBO {
  var sudokuData: GameDataBO
  do {
    sudokuData = GameDataBO()
    val (matrix, holes) = sudokuData
    matrix.generateMatrix()
    holes.digHolesByGameDifficulty(minEmpty, maxEmpty)
  } while (matrix.hasNotOnlyOneSolution(holes))
  return sudokuData
}

/**
 * 生成数独矩阵
 */
private fun List<MutableList<Int>>.generateMatrix() {
  //记录buildRandomArray()调用次数
  var currentTimes = 0
  var status = GenerateStatus.INIT_FIRST_ROW
  var row = 0
  var col = 0
  var tempRandomArray = mutableListOf<Int>()
  while (row != 9) {
    when (status) {
      GenerateStatus.FILL_GRID -> {
        if (currentTimes >= MAX_CALL_RANDOM_ARRAY_TIMES) {
          status = GenerateStatus.EMPTY_ALL
          continue
        }
        if (!isCandidateNmbFound(tempRandomArray, row, col)) {
          status = GenerateStatus.EMPTY_ROW
          continue
        }
        if (++col == 9) {
          row++
          status = GenerateStatus.FILL_ROW
        }
      }
      GenerateStatus.EMPTY_ROW -> {
        clearMatrixRow(row)
        status = GenerateStatus.FILL_ROW
      }
      GenerateStatus.FILL_ROW -> {
        tempRandomArray = buildRandomArray()
        currentTimes++
        col = 0
        status = GenerateStatus.FILL_GRID
      }
      GenerateStatus.INIT_FIRST_ROW -> {
        buildRandomArray().forEachIndexed { i, value -> this[0][i] = value }
        currentTimes = 1
        row = 0
        status = GenerateStatus.FILL_ROW
      }
      GenerateStatus.EMPTY_ALL -> {
        clearMatrix()
        status = GenerateStatus.INIT_FIRST_ROW
      }
    }
  }
}

/**
 * 生成数独矩阵算法的状态
 */
private enum class GenerateStatus {
  // 清空矩阵的所有数据
  EMPTY_ALL,

  // 清空矩阵的行数据
  EMPTY_ROW,

  // 填充矩阵的格子数据
  FILL_GRID,

  // 填充矩阵的行数据
  FILL_ROW,

  // 初始化矩阵的第一行数据
  INIT_FIRST_ROW
}

/**
 * 清空数独矩阵的行
 *
 * @param row    行
 */
private fun List<MutableList<Int>>.clearMatrixRow(row: Int) = this[row].fill(0)

/**
 * 清空数独矩阵
 */
private fun List<MutableList<Int>>.clearMatrix() = forEach { it.fill(0) }

/**
 * 尝试给数独矩阵的第row行第col列的数据赋值
 *
 * @param randomArray 随机数组
 * @param row         行
 * @param col         列
 * @return 能赋值返回true，不能赋值返回false
 */
private fun List<MutableList<Int>>.isCandidateNmbFound(randomArray: MutableList<Int>, row: Int, col: Int): Boolean {
  randomArray.filter { it != -1 }.forEachIndexed { i, random ->
    this[row][col] = random
    if (noConflict(row, col)) {
      randomArray[i] = -1;
      return true
    }
  }
  return false
}

/**
 * 判断该行、该列的数在数独矩阵中是否存在冲突
 *
 * @param row    行
 * @param col    列
 * @return 存在冲突返回false，不存在冲突返回true
 */
private fun List<List<Int>>.noConflict(row: Int, col: Int): Boolean {
  return this[row].checkRowIsOnly(col, col) && checkColumnIsOnly(row, col, row) && checkBlockIsOnly(row, col)
}

/**
 * 返回一个1-9的随机排列数组
 *
 * @return 打乱的数组
 */
private fun buildRandomArray(): MutableList<Int> {
  return (1..9).toMutableList().apply { shuffle() }
}

/**
 * 根据空缺格子数生成对应题目的空缺数组
 *
 * @param minEmpty 最小的空缺格子数
 * @param maxEmpty 最大的空缺格子数
 */
private fun List<MutableList<Boolean>>.digHolesByGameDifficulty(minEmpty: Int, maxEmpty: Int) {
  for (i in RandomUtil.randomInt(minEmpty, maxEmpty + 1) - 1 downTo 0) {
    this[i / 9][i % 9] = true
  }
  shuffle()
}
